File: proxy_server.py
#!/usr/bin/env python3
"""
Servidor proxy simple para evitar problemas de CORS con la API de Pokémon TCG
Ejecuta este script y luego usa http://localhost:8080 en lugar de la API directa
"""

from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import urllib.request
import json

class CORSProxyHandler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        """Manejar preflight requests de CORS"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'X-Api-Key, Content-Type')
        self.end_headers()
    
    def do_GET(self):
        """Proxificar solicitudes GET a la API de Pokémon TCG"""
        try:
            # Parsear la query string
            parsed_path = urlparse(self.path)
            query_params = parse_qs(parsed_path.query)
            
            # Construir URL de la API
            api_url = 'https://api.pokemontcg.io/v2/cards'
            if 'q' in query_params:
                api_url += f'?q={query_params["q"][0]}'
            
            # Crear solicitud a la API
            req = urllib.request.Request(api_url)
            
            # Pasar el header X-Api-Key si está presente
            if 'X-Api-Key' in self.headers:
                req.add_header('X-Api-Key', self.headers['X-Api-Key'])
            
            # Realizar solicitud
            with urllib.request.urlopen(req) as response:
                data = response.read()
                status_code = response.getcode()
                
                # Enviar respuesta con headers CORS
                self.send_response(status_code)
                self.send_header('Access-Control-Allow-Origin', '*')
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(data)
                
        except Exception as e:
            self.send_response(500)
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            error_response = json.dumps({'error': str(e)}).encode()
            self.wfile.write(error_response)
    
    def log_message(self, format, *args):
        """Suprimir logs del servidor"""
        pass

def run(server_class=HTTPServer, handler_class=CORSProxyHandler, port=8080):
    server_address = ('', port)
    httpd = server_class(server_address, handler_class)
    print(f'🚀 Servidor proxy CORS ejecutándose en http://localhost:{port}')
    print(f'📡 Proxificando solicitudes a https://api.pokemontcg.io/v2/cards')
    print('Presiona Ctrl+C para detener el servidor')
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print('\n🛑 Deteniendo servidor...')
        httpd.shutdown()

if __name__ == '__main__':
    run()

File: proxy_server.js
/**
 * Servidor proxy simple para evitar problemas de CORS con la API de Pokémon TCG
 * Ejecuta: node proxy_server.js
 * Luego usa http://localhost:8080 en lugar de la API directa
 */

const http = require('http');
const https = require('https');
const url = require('url');

const PORT = 8080;

const server = http.createServer((req, res) => {
    // Manejar CORS preflight
    if (req.method === 'OPTIONS') {
        res.writeHead(200, {
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
            'Access-Control-Allow-Headers': 'X-Api-Key, Content-Type',
            'Access-Control-Max-Age': '86400'
        });
        res.end();
        return;
    }

    // Solo manejar GET requests
    if (req.method !== 'GET') {
        res.writeHead(405, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ error: 'Method not allowed' }));
        return;
    }

    try {
        // Parsear query parameters
        const parsedUrl = url.parse(req.url, true);
        const query = parsedUrl.query.q || '';
        
        // Construir URL de la API
        const apiUrl = `https://api.pokemontcg.io/v2/cards?q=${query}`;
        
        // Obtener API key del header
        const apiKey = req.headers['x-api-key'] || '';
        
        console.log(`📡 Proxificando: ${apiUrl}`);
        
        // Realizar solicitud a la API
        const options = {
            headers: {
                'X-Api-Key': apiKey,
                'Accept': 'application/json'
            }
        };
        
        https.get(apiUrl, options, (apiRes) => {
            let data = '';
            
            apiRes.on('data', (chunk) => {
                data += chunk;
            });
            
            apiRes.on('end', () => {
                // Enviar respuesta con headers CORS
                res.writeHead(apiRes.statusCode, {
                    'Access-Control-Allow-Origin': '*',
                    'Content-Type': 'application/json'
                });
                res.end(data);
            });
        }).on('error', (error) => {
            console.error('❌ Error:', error.message);
            res.writeHead(500, {
                'Access-Control-Allow-Origin': '*',
                'Content-Type': 'application/json'
            });
            res.end(JSON.stringify({ error: error.message }));
        });
        
    } catch (error) {
        res.writeHead(500, {
            'Access-Control-Allow-Origin': '*',
            'Content-Type': 'application/json'
        });
        res.end(JSON.stringify({ error: error.message }));
    }
});

server.listen(PORT, () => {
    console.log('🚀 Servidor proxy CORS ejecutándose en http://localhost:' + PORT);
    console.log('📡 Proxificando solicitudes a https://api.pokemontcg.io/v2/cards');
    console.log('Presiona Ctrl+C para detener el servidor');
});

